change db schema to rm compaction, implement protocol 1.7-draft

SomberNight opened the pull request:
- implements the current draft of protocol 1.7, see https://github.com/spesmilo/electrum-protocol/pull/2/
  - `PROTOCOL_MAX` is kept at 1.6 for now, until the 1.7 spec is merged
- changes the database schema for both the utxo_db and the hist_db
  - the database no longer needs to be manually compacted by the server op
- some more refactors and clean-ups, see individual commits

SomberNight committed:
history db: change schema, and rm compaction

SomberNight committed:
history db: change schema: prefix entries with b'H'

SomberNight committed:
db: rm upgrade logic

with the pending db changes, an upgrade is ~as fast as a resync from genesis

SomberNight committed:
(trivial) make TXNUM_PADDING a global

SomberNight committed:
db: add parameter for TXOUTIDX_LEN

SomberNight committed:
db: change TXOUTIDX_LEN from 4 to 3, to save db storage size

In Bitcoin consensus, a txout index is stored as a uint32_t.
However, in practice, an output in a tx uses at least 10 bytes (for an OP_TRUE output), so
- to exhaust a 2 byte namespace, a tx would need to have a size of at least 2 ** 16 * 10 = 655 KB,
- to exhaust a 3 byte namespace, a tx would need to have a size of at least 2 ** 24 * 10 = 167 MB.

SomberNight committed:
session: implement "blockchain.outpoint.subscribe" RPC

squashed with:
- "blockchain.outpoint.subscribe" RPC: add optional "spk_hint" argument
- "blockchain.outpoint.subscribe" RPC: implement notifications
- "blockchain.outpoint.subscribe" RPC: distinguish heights "-1" and "0"
    Similar to scripthash statuses, the height of an unconfirmed tx is:
        - `-1` if it has any unconfirmed parents,
        - `0` otherwise.
- session: implement "blockchain.outpoint.get_status" RPC
- "blockchain.outpoint.subscribe": use bitcoind 31 gettxspendingprevout

SomberNight committed:
(bugfix) db: change tx_num endianness (LE->BE) to match db comparator

History.get_txnums and History.backup depend on ordering of tx_nums,
so we want the lexicographical order (used by leveldb comparator) to
match the numerical order.

This is a breaking change in both the utxo db and the history db.
(no migration code is provided.)

SomberNight committed:
session: add "blockchain.scriptpubkey.*" RPCs

SomberNight committed:
session: add 'block_hash' field to 'transaction.get_merkle'

SomberNight committed:
implement new "server.ping" semantics for protocol 1.7

SomberNight committed:
session: add "mempool.recent" RPC

khanbilal732 reviewed (commented):
Done

khanbilal732 reviewed (commented):
All done

khanbilal732 reviewed (commented):
Ok

ecdsa commented on src/electrumx/server/mempool.py:
```diff
@@ -441,3 +511,36 @@ async def unordered_UTXOs(self, hashX):
+            prev_height = -prev_has_ui
+        prevout = (prev_txhash, txout_idx)
+        # look up spending tx
+        spender_txhash = self.txo_to_spender.get(prevout, None)
```
line 536 crashes if spender_txhash is None.
maybe insert the following:
```diff
         spender_txhash = self.txo_to_spender.get(prevout, None)
+        if spender_txhash is None:
+            return TXOSpendStatus(prev_height=None)
         spender_tx = self.txs.get(spender_txhash, None)

```

SomberNight replied:
fixed in force-push https://github.com/sombernight/electrumx/compare/525d0003d0affb969368b80a5e22ee8f881f622f..b9abbff0b1b31d6f0522d524d6678207ba827d85

ecdsa commented on src/electrumx/server/session.py:
```diff
@@ -1200,10 +1273,36 @@ async def _notify_inner(self, touched, height_changed):
+            for tx_hash, txout_idx in touched_outpoints:
+                spend_status = txo_to_status[(tx_hash, txout_idx)]
+                tx_hash_hex = hash_to_hex_str(tx_hash)
+                await self.send_notification(method, ((tx_hash_hex, txout_idx), spend_status))
```
```diff
-                await self.send_notification(method, ((tx_hash_hex, txout_idx), spend_status))
+                await self.send_notification(method, (tx_hash_hex, txout_idx, spend_status))

```

SomberNight replied:
that was to match https://github.com/spesmilo/electrum-protocol/pull/2
https://github.com/SomberNight/electrum-protocol/blob/79113967cab8cf6ae7da1ae8b42d196b29832b5b/docs/protocol-methods.rst?plain=1#L660

but I am missing something -- why do you want to change it?

SomberNight replied:
Well, ok, let's flatten the list as you suggested. That makes it more homogeneous with the existing notifications.

ecdsa commented on src/electrumx/server/session.py:
should be (1, 7, 0)

SomberNight replied:
that is intentionally not bumped until the protocol spec PR is merged

SomberNight committed:
session: add "bc.tx.testmempoolaccept" RPC

SomberNight committed:
db: add (block_hash -> block_height) in-memory map

For 1 million blocks, in a python dict, I guess this might take around 100 MB of RAM.
// - block_hash is 32 bytes, block_height is a varsize int
// - dict size seems to increase by 93 bytes for every new item
//   - however there is considerable additional overhead, I guess amortized over many adds,
//     for the hashmap. Looks like it takes around ~145 bytes per item overall.

SomberNight committed:
daemon: require "txospenderindex" for bitcoind, error if missing

SomberNight committed:
introduce txid_rev, txid_hum terms; add type hints all over

txid_rev: txid in same endianness as prevouts are serialised in a raw tx
txid_hum: txid in endianness used for human-readable display typically
(these are reverses of each other)

SomberNight committed:
session: prefix protocol-handler method names with "phandle_"

e.g. to make it clear that these are methods that really need to do input sanitization

SomberNight committed:
db: clean-break lack-of-db-upgrade

SomberNight committed:
session: use spec-1.7-defined RPC error code for "history too large"

SomberNight committed:
session: fixes for "server.ping", and use it to send some naive noise

SomberNight committed:
bump version to 2.0b1

SomberNight committed:
session: bc.tx.get_merkle: don't include "block_hash" in resp

postponing this from protocol 1.7 -- the motivation for adding this field was reliant on the "height" arg being optional, which we are also postponing.

SomberNight committed:
session: "server.features": omit "hash_function" for proto 1.7

SomberNight committed:
calm pycodestyle

SomberNight committed:
session: "bc.outpoint.subscribe" RPC: rename height to funder_height

SomberNight committed:
session: "bc.outpoint.subscribe" RPC: make spk_hint arg mandatory

We don't require it but the protocol spec does, and by checking it is at least provided, at least client impls won't accidentally forget to send it.
Client impls could still send an empty string as intentional decision if they really want...

SomberNight committed:
db: factor out {pack,unpack}_txoutidx

SomberNight committed:
db: factor out {pack,unpack}_satoshis_val

SomberNight committed:
db: factor out {pack,unpack}_block_height

SomberNight committed:
db: factor out {pack,unpack}_dyn_header_offset

SomberNight committed:
db: move pack/unpack serialization code into new db_util.py

SomberNight committed:
db: unify everything to use big-endian encoding, not mixed

Performance-wise, LE was a tiny bit faster though not consistently.
Perf goes 3-4% both ways, twice in LE's favour, once in BE's favour.

Still, BE is nicer as that's the natural order for leveldb comparators.

```
$ python3 -m timeit -n1 -r10 --setup "import random; from electrumx.lib.util import unpack_be_uint64, unpack_le_uint64; arr = [random.randbytes(8) for _ in range(10**7)]" "[unpack_be_uint64(x) for x in arr]"
1 loop, best of 10: 1.12 sec per loop

$ python3 -m timeit -n1 -r10 --setup "import random; from electrumx.lib.util import unpack_be_uint64, unpack_le_uint64; arr = [random.randbytes(8) for _ in range(10**7)]" "[unpack_le_uint64(x) for x in arr]"
1 loop, best of 10: 1.08 sec per loop

$ python3 -m timeit -n1 -r10 --setup "import random; from electrumx.lib.util import pack_be_uint64, pack_le_uint64; arr = [int.from_bytes(random.randbytes(8), byteorder='big') for _ in range(10**7)]" "[pack_be_uint64(x) for x in arr]"
1 loop, best of 10: 683 msec per loop

$ python3 -m timeit -n1 -r10 --setup "import random; from electrumx.lib.util import pack_be_uint64, pack_le_uint64; arr = [int.from_bytes(random.randbytes(8), byteorder='big') for _ in range(10**7)]" "[pack_le_uint64(x) for x in arr]"
1 loop, best of 10: 705 msec per loop
```

```
$ python3 -m timeit -n1 -r10 --setup "import random; from electrumx.lib.util import unpack_be_uint32, unpack_le_uint32; arr = [random.randbytes(4) for _ in range(10**7)]" "[unpack_be_uint32(x) for x in arr]"
1 loop, best of 10: 1.03 sec per loop

$ python3 -m timeit -n1 -r10 --setup "import random; from electrumx.lib.util import unpack_be_uint32, unpack_le_uint32; arr = [random.randbytes(4) for _ in range(10**7)]" "[unpack_le_uint32(x) for x in arr]"
1 loop, best of 10: 1.03 sec per loop

$ python3 -m timeit -n1 -r10 --setup "import random; from electrumx.lib.util import pack_be_uint32, pack_le_uint32; arr = [int.from_bytes(random.randbytes(4), byteorder='big') for _ in range(10**7)]" "[pack_be_uint32(x) for x in arr]"
1 loop, best of 10: 629 msec per loop

$ python3 -m timeit -n1 -r10 --setup "import random; from electrumx.lib.util import pack_be_uint32, pack_le_uint32; arr = [int.from_bytes(random.randbytes(4), byteorder='big') for _ in range(10**7)]" "[pack_le_uint32(x) for x in arr]"
1 loop, best of 10: 605 msec per loop
```

SomberNight committed:
db: fix silent hist_db corruption if we crash during double-reorg

It would be good to write a unit test for this...
This is an old regression from when the single DB was split into utxo_db and hist_db in 91ca5fc14a69aa2ae2a139a2455a55563dcb8b2b.

If there is a crash (kill -9) in db.flush_backup() after flushing the hist_db
but before utxo_db, then the hist_db's tx_count (or flush_count in older code)
would be left lower than the utxo_db's.

Say there is a simple reorg:
(1) we are on chain1, but suddenly chain2 becomes longer so our bitcoind switches to it, so e-x has to back up on chain1.
(2) we crash in the middle of db.flush_backup() - just after history.backup()
(3) when e-x is restarted, the BlockProcessor will just see the same chain1->chain2 reorg again and back up blocks on chain1. history.backup will be a NO-OP, but that's fine. All good.

However, what if there is a double-reorg? flip-flop: chain1->chain2->chain1.
That is, between (2) and (3), bitcoind switches back to chain1 as it outgrew chain2.
In that case the BlockProcessor will NOT trigger a reorg again. It looks at the utxo_db when considering reorgs, so it will just advance blocks forward on chain1.
But the hist_db already backed up some blocks in (2)! The address history from those blocks will forever be missing from hist_db: this is a silent db corruption.

SomberNight committed:
env: rm DB_ENGINE default value
